<h1 id="amplicon-tutorial-for-ideelers">Amplicon Tutorial for IDEElers</h1>
<h2 id="session-4-variant-discovery">Session 4: Variant Discovery</h2>
<p>Here we will explore basic alignment strategies.</p>
<h3 id="resources">Resources</h3>
<ol style="list-style-type: decimal">
<li>Read in its entirety -- including frequently asked questions and review each ppt available -- GATK's Best Practices for <a href="https://software.broadinstitute.org/gatk/best-practices/bp_3step.php?case=GermShortWGS&amp;p=2">Germline Muations in WGS and Exome Sequences, Variant Discovery Page</a> and <a href="https://software.broadinstitute.org/gatk/best-practices/bp_3step.php?case=GermShortWGS&amp;p=3">Germline Muations in WGS and Exome Sequences, Callset Refinement</a></li>
<li><strong>Read in its entirety</strong> <a href="https://gatkforums.broadinstitute.org/gatk/discussion/1268/what-is-a-vcf-and-how-should-i-interpret-it">What is a VCF and how should I interpret it? by GATK's Geraldine</a></li>
</ol>
<h3 id="exercises">Exercises</h3>
<ol style="list-style-type: decimal">
<li><p>Finish the simulation from the Rmarkdown file <code>ToyDataGenerator.Rmd</code>. Explore the VCF that you created. Does it produce the mutation that you introduced in the <code>ToyDataGenerator.Rmd</code>? Introduce more mutations or change the frequency at which you expect to see the mutatin by concatenating several different fastas together in R or on the command line.</p></li>
<li><p>Load the VCF file into R using the R package <code>VCFR</code>, you can dowload it following the directions <a href="https://github.com/knausb/vcfR">here</a>. Manipulate the VCF and look at GT and depth of coverage. Complete the <code>VCFR</code> tutorials (i.e Vignettes) located <a href="https://cran.r-project.org/web/packages/vcfR/index.html">here</a>. <strong>This will help provide an issue and manageable way to refine your VCF discoveries to higher quality SNPs.</strong></p></li>
</ol>
