<h1 id="amplicon-tutorial-for-ideelers">Amplicon Tutorial for IDEElers</h1>
<h2 id="session-3-alignments">Session 3: Alignments</h2>
<p>Here we will explore basic alignment strategies.</p>
<h3 id="resources">Resources</h3>
<ol style="list-style-type: decimal">
<li>Read in its entirety -- including frequently asked questions and review each ppt available -- GATK's Best Practices for <a href="https://software.broadinstitute.org/gatk/best-practices/bp_3step.php?case=GermShortWGS&amp;p=1">Germline Muations in WGS and Exome Sequences, Pre-Processing Page</a></li>
<li>Review the articles and manuals for the most commonly used aligners: <a href="http://bio-bwa.sourceforge.net/">bwa-mem</a> and <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie2</a></li>
<li>Review the format of a <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">Sam File</a></li>
</ol>
<h3 id="exercises">Exercises</h3>
<p>Simulate Fasta and align it to a referent genome following the Rmarkdown file <code>ToyDataGenerator.Rmd</code>. Make sure to change the probability of the A,C,T,G calls and see how it affects the fasta file. Open the fasta file with a program like <code>Mega</code> or <code>AliView</code>. Paste several fasta files together and perform multiple pairwise alignment using <code>Muscle</code> or <code>CLUSTALW</code> (preloaded in the previously mentioned programs or in the R package <code>Ape</code>.</p>
